Improve call provenance during GraphModule scripting

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50538 Improve call provenance during GraphModule scripting**

Differential Revision: [D25935403](https://our.internmc.facebook.com/intern/diff/D25935403)

Commit:
Improve call provenance during GraphModule scripting

[ghstack-poisoned]

Commit:
Update on "Improve call provenance during GraphModule scripting"




[ghstack-poisoned]

Commit:
Update on "Improve call provenance during GraphModule scripting"




[ghstack-poisoned]

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -137,6 +138,18 @@ def __new__(cls: 'Type[GraphModule]', *args, **kwargs):
+        parent_frame = inspect.currentframe().f_back
+        calling_name = cls.__name__
+        if "root" in parent_frame.f_locals:
+            calling_name = parent_frame.f_locals["root"].__class__.__name__
```
This is too brittle; It's making assumptions about the caller and breaks the procedural abstraction. Callees should not assume anything about the caller beyond the contract established by the function signature

I think we should instead do the following:

1. Here in `__new__`, set `GraphModuleImpl.__name__ = 'GraphModule'`. This will be the default value
2. Add a parameter `class_name : Optional[str] = None` to the `GraphModule` constructor. Internally, the constructor will set `self.__class__.__name__ = class_name` if that name is not None
3. `symbolic_trace` will pass the value of `root.__class__.__name__` for this parameter

Note that the case where we want to populate the `class_name` parameter is limited to `symbolic_trace`. In other instances of creating a `GraphModule`, it's the caller's (i.e. some pass writer's) responsibility to provide a name that makes sense

Commit:
Update on "Improve call provenance during GraphModule scripting"




[ghstack-poisoned]

Commit:
Update on "Improve call provenance during GraphModule scripting"




[ghstack-poisoned]

jamesr66a commented on torch/fx/symbolic_trace.py:
```diff
@@ -449,4 +449,6 @@ def symbolic_trace(root : Union[torch.nn.Module, Callable]) -> GraphModule:
-    return GraphModule(root if isinstance(root, torch.nn.Module) else torch.nn.Module(), Tracer().trace(root))
+    return GraphModule(root if isinstance(root, torch.nn.Module) else torch.nn.Module(),
+                       Tracer().trace(root),
+                       root.__class__.__name__)
```
In the case that `root` is a function, I think it makes more sense to pass `root.__name__` here:

```

def my_cool_fn(x):
    return x

print(my_cool_fn.__class__.__name__)
"""
function
"""

print(my_cool_fn.__name__)
"""
my_cool_fn
"""
```

jamesr66a commented on test/test_fx.py:
```diff
@@ -1263,5 +1268,28 @@ def forward(self, x: List[str]) -> List[str]:
+
+        with self.assertRaisesRegex(RuntimeError, "'wrapper_fn' is "
+                                    "being compiled since it was called"
+                                    " from 'function.forward'"):
```
The user may be confused that we're saying `<fn_name>.forward` but not sure of a quick solution to this, since we're promoting functions to Modules in symbolic tracing

"""All abstractions are leaky abstractions when it comes to error reporting" - Me" - Wayne Gretzky" - Michael Scott

ansley replied:
Same here--agree that it's confusing, not sure what to do about it. An earlier version of my code set the module name to `nnModule_with_autogenerated_forward_wrapping_user_function_FUNCTION-NAME`, which is clearer, but also way too long. I'll keep it as-is unless you think renaming the module would be better.

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -156,8 +156,13 @@ def __init__(self, root: Union[torch.nn.Module, Dict[str, Any]], graph: Graph):
 
+            name (str): ``str`` denotes the name of this GraphModule for debugging purposes. If it's unset, all
+                error messages will report as originating from ``GraphModule``; it may be helpful to set this to
+                ``root``'s original name.
```
"it may be helpful to set this to ``root``'s original name or a name that makes sense within the context of your transform"

jamesr66a reviewed (commented):
Looking good! Some small stuff

Commit:
Update on "Improve call provenance during GraphModule scripting"




[ghstack-poisoned]

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -156,8 +156,13 @@ def __init__(self, root: Union[torch.nn.Module, Dict[str, Any]], graph: Graph):
 
             graph (Graph): ``graph`` contains the nodes this GraphModule should use for code generation
 
+            name (str): ``str`` denotes the name of this GraphModule for debugging purposes. If it's unset, all
```
I think this should say either "str that denotes..." or "name denotes..."

jamesr66a reviewed (approved):
![image](https://user-images.githubusercontent.com/4685384/104787163-bbcafd80-5743-11eb-81a0-13c0871ca465.png)

Commit:
Update on "Improve call provenance during GraphModule scripting"




[ghstack-poisoned]

Commit:
Update on "Improve call provenance during GraphModule scripting"


Differential Revision: [D25935403](https://our.internmc.facebook.com/intern/diff/D25935403)

[ghstack-poisoned]

Commit:
Update on "Improve call provenance during GraphModule scripting"


Differential Revision: [D25935403](https://our.internmc.facebook.com/intern/diff/D25935403)

[ghstack-poisoned]
